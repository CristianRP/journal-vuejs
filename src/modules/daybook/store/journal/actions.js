import journalApi from '@/api/journalApi'
// export const myAction = async({ commit }) => {


// }

export const loadEntries = async({ commit }) => {
  const { data } = await journalApi.get('/entries.json')

  if ( !data ) {
    commit('setEntries', [])
  }

  const entries = []
  for (let id of Object.keys(data)) {
    entries.push({
      id,
      ...data[id]
    })
  }

  commit('setEntries', entries)
}

export const updateEntry = async({ commit }, entry) => { // entry should be a param

  // exctract only what we need // -id
  const { date, text } = entry
  const dataToSave = { date, text}

  // await journalApi.put (path, .json, dataToSave)
  await journalApi.put(`/entries/${entry.id}.json`, dataToSave)

  // Commit a mutation
  commit('updateEntry', entry)
}

export const createEntry = async({ commit }, entry) => {
  const { date, text } = entry
  const dataToSave = { date, text }

  const { data } = await journalApi.post(`/entries.json`, dataToSave)

  dataToSave.id = data.name

  commit('addEntry', dataToSave)

  return data.name
}

export const deleteEntry = async({ commit }, id) => {

  await journalApi.delete(`/entries/${id}.json`)

  commit('deleteEntry', id)

  return id
}
