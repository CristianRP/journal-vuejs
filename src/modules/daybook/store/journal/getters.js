// export const myGetter = (state) => {
  // return state
// }

export const getEntrieByTerm = (/* state */) => {
  // return state
}

export const getEntryById = (/* state */) => {
  // return state
}


