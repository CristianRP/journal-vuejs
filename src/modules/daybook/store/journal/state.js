export default () => ({
  isLoading: true,
  entries: [
    {
      id: new Date().getTime(),
      date: new Date().toDateString(),
      text: 'Labore dolor laborum aliqua est excepteur labore Lorem. Laborum sint elit sit culpa labore. Eiusmod id et amet elit exercitation. Eu amet deserunt commodo voluptate et. Consequat cupidatat aliqua id eu tempor mollit laborum sunt cupidatat ipsum Lorem. Reprehenderit labore do amet proident adipisicing nisi exercitation. Consectetur do magna nulla ut eiusmod do nulla veniam officia laborum et cillum tempor.',
      picture: null, // https:
    },
    {
      id: new Date().getTime() + 1000,
      date: new Date().toDateString(),
      text: 'Elit reprehenderit sit voluptate velit proident non labore pariatur dolore. Nisi aute amet adipisicing sint. Magna commodo minim ea reprehenderit minim cupidatat. Quis dolore incididunt reprehenderit enim.',
      picture: null, // https:
    },
    {
      id: new Date().getTime() + 2000,
      date: new Date().toDateString(),
      text: 'Veniam nisi ipsum laborum cupidatat eu ea. Elit et sit aute aute laborum incididunt incididunt fugiat ullamco labore eu incididunt. Lorem aliqua ad id sunt cillum excepteur consequat. Qui elit qui veniam elit ex.',
      picture: null, // https:
    }
  ]
})
