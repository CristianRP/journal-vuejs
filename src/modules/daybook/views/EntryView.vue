<template>
  <template v-if="entry">
      <div class="entry-title d-flex justify-content-between p-2">
      <div>
        <span class="text-success fs-3 fw-bold">{{ day }}</span>
        <span class="mx-1 fs-3">{{ month }}</span>
        <span class="mx-2 fs-4 fw-light">{{ yearDay }}</span>
      </div>

      <div>

        <input
          type="file"
          ref="imageSelector"
          v-show="false"
          accept="image/png, image/jpeg"
          @change="onSelectedImage">

        <button
          v-if="entry.id"
          @click="onDeleteEntry"
          class="btn btn-danger mx-2">
          Delete
          <i class="fa fa-trash-alt"></i>
        </button>
        <button class="btn btn-primary" @click="onSelectImage">
          Upload Photo
          <i class="fa fa-upload"></i>
        </button>
      </div>
    </div>
    <hr>
    <!-- Need to add a template to wrap both root -->
    <div class="d-flex flex-column px-3 h-75">
      <textarea placeholder="What happened today?" v-model="entry.text"></textarea>

      <FabButton
        icon="fa-save"
        @on:click="saveEntry" />

      <img
        v-if="entry.picture && !localImage"
        :src="entry.picture"
        alt="entry-picture"
        class="img-thumbnail">

      <img
        v-if="localImage"
        :src="localImage"
        alt="entry-picture"
        class="img-thumbnail">
    </div>
  </template>
</template>

<script>
import { defineAsyncComponent } from 'vue'
import { mapGetters, mapActions } from 'vuex'
import Swal from 'sweetalert2'

import getDayMonthYear from '../helpers/getDayMonthYear'
import uploadImage from '../helpers/uploadImage'

export default {
  name: 'EntryView',
  props: {
    id: { type: String, required: true }
  },
  data() {
    return {
      entry: null,
      localImage: null,
      file: null
    }
  },
  components: {
    FabButton: defineAsyncComponent(() => import('../components/FabButton'))
  },
  created() {
    this.loadEntry()
  },
  mounted() {
    // this.loadEntry()
  },
  computed: {
    ...mapGetters('journal', ['getEntryById']),
    shortText() {
      return ( this.entry.text.length > 130 )
        ? `${this.entry.text.substring(0, 130)}...`
        : this.entry.text
    },
    day() {
      const { day } = getDayMonthYear( this.entry.date )
      return day
    },
    month() {
      const { month } = getDayMonthYear( this.entry.date )
      return month
    },
    yearDay() {
      const { yearDay } = getDayMonthYear( this.entry.date )
      return yearDay
    }
  },
  methods: {
    loadEntry() {
      let entry;
      if ( this.id === 'new' ) {
        entry = {
          text: '',
          date: new Date().getTime()
        }
      } else {
        entry = this.getEntryById(this.id)
        if ( !entry ) return this.$router.push({ name: 'no-entry' })
      }

      this.entry = entry
    },

    async saveEntry() {

      Swal.fire({
        title: 'Please wait',
        allowOutsideClick: false
      })
      Swal.showLoading()

      const picture = await uploadImage( this.file )

      this.entry.picture = picture

      if (this.entry.id) {
        await this.updateEntry(this.entry)
      } else {
        const id = await this.createEntry(this.entry)

        this.$router.push({ name: 'entry', params: { id } })
      }

      this.file = null
      this.localImage = null
      Swal.fire('Saved', 'Entry saved properly', 'success')
    },

    async onDeleteEntry() {

      const { isConfirmed } = await Swal.fire({
        title: 'Are you sure?',
        text: 'This will erase the entry permanently',
        showDenyButton: true,
        confirmButtonText: 'Yes, delete',
        denyButtonText: 'Cancel'
      })

      if (isConfirmed) {
        Swal.fire({
          title: 'Please wait',
          allowOutsideClick: false
        })
        Swal.showLoading()

        await this.deleteEntry( this.entry.id )

        this.$router.push({ name: 'no-entry' })

        Swal.fire('Deleted', '', 'success')
      }
    },

    onSelectedImage( event ) {
      const file = event.target.files[0]

      if ( !file ) {
        this.localImage = null
        this.file = null
        return
      }

      this.file = file

      const fr = new FileReader()
      fr.onload = () => this.localImage = fr.result
      fr.readAsDataURL( file )
    },

    onSelectImage() {
      this.$refs.imageSelector.click()
    },

    ...mapActions('journal', ['updateEntry', 'createEntry', 'deleteEntry'])
  },
  watch: {
    // eslint-disable-next-line no-unused-vars
    id() {
      this.loadEntry()
    }
  }
}
</script>

<style lang="scss" scoped>
textarea {
  font-size: 20px;
  border: none;
  height: 100%;

  &:focus {
    outline: none;
  }
}

img {
  width: 200px;
  position: fixed;
  bottom: 150px;
  right: 20px;
  box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
}
</style>
