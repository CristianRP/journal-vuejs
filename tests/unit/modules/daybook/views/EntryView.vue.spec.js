import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'

import journal from '@/modules/daybook/store/journal'
import { journalState } from '../../../mock-data/test-journal-state'

import EntryView from '@/modules/daybook/views/EntryView'

import Swal from 'sweetalert2'

const createVuexStore = ( initialStore ) => createStore({
  modules: {
    journal: {
      ...journal,
      state: { ...initialStore }
    }
  }
})

jest.mock('sweetalert2', () => ({
  fire: jest.fn(),
  showLoading: jest.fn(),
  close: jest.fn()
}))

describe('Tests on EntryView component', () => {
  let wrapper

  const store = createVuexStore( journalState )
  store.dispatch = jest.fn()

  const mockRouter = {
    push: jest.fn()
  }

  beforeEach(() => {
    jest.clearAllMocks()

    wrapper = shallowMount( EntryView, {
      props: {
        id: '-N3PLntxWSjEgGDGZepB'
      },
      global: {
        mocks: {
          $router: mockRouter
        },
        plugins: [ store ]
      }
    })
  })

  test('should return the user to home if id does not exists', () => {
    wrapper = shallowMount( EntryView, {
      props: {
        id: 'does_not_exists'
      },
      global: {
        mocks: {
          $router: mockRouter
        },
        plugins: [ store ]
      }
    })

    expect( mockRouter.push ).toHaveBeenCalled()
    expect( mockRouter.push ).toHaveBeenCalledWith({ name: 'no-entry' })
  })

  test('should show the proper entry', () => {
    expect( wrapper.html() ).toMatchSnapshot()
    expect( mockRouter.push ).not.toHaveBeenCalled()
  })

  test('should delete the entry', (done) => {
    Swal.fire.mockReturnValueOnce( Promise.resolve({ isConfirmed: true }))

    wrapper.find('.btn-danger').trigger('click')

    expect( Swal.fire ).toHaveBeenCalledWith({
      title: 'Are you sure?',
      text: 'This will erase the entry permanently',
      showDenyButton: true,
      confirmButtonText: 'Yes, delete',
      denyButtonText: 'Cancel'
    })

    setTimeout(() => {
      expect( store.dispatch ).toHaveBeenCalledWith("journal/deleteEntry", "-N3PLntxWSjEgGDGZepB")
      expect( mockRouter.push ).toHaveBeenCalled()
      done()
    }, 1);
  })
})
