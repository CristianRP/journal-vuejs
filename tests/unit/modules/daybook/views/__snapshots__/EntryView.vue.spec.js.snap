// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on EntryView component should show the proper entry 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">31</span><span class="mx-1 fs-3">May</span><span class="mx-2 fs-4 fw-light">2022, Thursday</span></div>
  <div><input type="file" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Delete <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Upload Photo <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<!-- Need to add a template to wrap both root -->
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="What happened today?"></textarea>
  <fab-button-stub icon="fa-save"></fab-button-stub>
  <!--v-if-->
  <!--v-if-->
</div>
`;
